src/common/src/array/list_array.rs

```diff
@@ -624,6 +626,230 @@ impl<'a> From<&'a ListValue> for ListRef<'a> {
+                        (_, '"') => return Err("Unexpected array element.".to_string()),
+                        _ => {}
+                    }
+                };
+                Ok(Some(ScalarImpl::from_literal(&s, self.data_type)?))
+            }
+
+            /// Parse a double quoted non-array value.
+            fn parse_quoted(&mut self) -> Result<ScalarImpl, String> {
+                assert!(self.try_consume('"'));
+                // peek until the next unescaped '"'
+                let mut chars = self.input.char_indices();
+                let mut has_escape = false;
+                let s = loop {
+                    match chars.next().ok_or_else(Self::eoi)? {
+                        (_, '\\') => {
+                            has_escape = true;
+                            chars.next().ok_or_else(Self::eoi)?;
+                        }
+                        (i, '"') => {
+                            let s = &self.input[..i];
+                            self.input = &self.input[i + 1..];
+                            break if has_escape {
+                                Cow::Owned(Self::unescape(s))
+                            } else {
+                                Cow::Borrowed(s)
+                            };
+                        }
+                        _ => {}
+                    }
+                };
+                ScalarImpl::from_literal(&s, self.data_type)
+            }
+
+            /// Unescape a string.
+            fn unescape(s: &str) -> String {
+                let mut unescaped = String::with_capacity(s.len());
+                let mut chars = s.chars();
+                while let Some(mut c) = chars.next() {
+                    if c == '\\' {
+                        c = chars.next().unwrap();
+                    }
+                    unescaped.push(c);
+                }
+                unescaped
+            }
+
+            /// Unescape a string and trim the trailing whitespaces.
+            ///
+            /// Example: `"\  " -> " "`
+            fn unescape_trim_end(s: &str) -> String {
+                let mut unescaped = String::with_capacity(s.len());
+                let mut chars = s.chars();
+                let mut len_after_last_escaped_char = 0;
+                while let Some(mut c) = chars.next() {
+                    if c == '\\' {
+                        c = chars.next().unwrap();
+                        unescaped.push(c);
+                        len_after_last_escaped_char = unescaped.len();
+                    } else {
+                        unescaped.push(c);
+                    }
+                }
+                let l = unescaped[len_after_last_escaped_char..].trim_end().len();
+                unescaped.truncate(len_after_last_escaped_char + l);
+                unescaped
+            }
+
+            /// Consume the next 4 characters if it matches "null".
+            fn try_parse_null(&mut self) -> bool {
```

Comment:
Given this is only called by `self.data_type.is_array()` now, it can be simplified (or be more specific to prevent misuse). Some ideas include:
* It is an error if not started with `{` or `null`.
* After `null` and `skip_whitespace`, the next char must be `,` or `}`.

^ Not saying we should implement these checks here, but that we can leverage these properties here.

Reply:
Even though it is only used by the array branch, I don't think this function should do anything other than what its name suggests, e.g. checking for leading or following chars. These should be left to the upper-level function. 🤔

Reply:
As I said, I am not saying we should implement these checks here, but that we can leverage these properties here.

Let me try to rephrase issues I want to address here:
* When peeking the next char, `{` shall not be allowed.
* When peeking the next char, it is better to reuse `skip_whitespace` rather than listing those chars again.
* We have fixed a bug by not calling this function when parsing non-array values. However, its doc / name / implementation still looks like very generic and usable there. Although we can prevent this misuse by having unit test, it is better to make it clear from the code logic - that this function only handles limited case correctly.

Reply:
> it is better to make it clear from the code logic

To clarify, I do not have an idea of how to achieve this yet. So the suggestions here sound vague. The easiest is just to leave some comments, clearly stating this will cause bug if used when parsing non-array value, and point to the corresponding unit test case as an example.

Reply:
Sorry again for my misunderstanding. My intention is to make this function generic and reusable. But it is only useful in parsing arrays. I've added a note to make it clear.